Added section for get all destinations (#193)
SAP/cloud-sdk: docs/java/features/connectivity/destinations.md

This concept is integrated with the [Destination Service](https://help.sap.com/viewer/cca91383641e40ffbe03bdc78f00f681/Cloud/en-US/7e306250e08340f89d6c103e28840f30.html) that is available on SAP Cloud Platform. If the application has a service binding to this service in place the SDK will provide access to these destinations.

> ## Accessing All Destinations in Cloud Foundry ##

Reviewer: ```suggestion
## Get a List of all Destination in Cloud Foundry  ##
```

Author: yup your suggestion is better :).

> ## Get a List of all Destination in Cloud Foundry  ##

Reviewer 2: ```suggestion
## Get a List of all Destination from Destination Service on Cloud Foundry  ##
```

>  Inorder to fetch all destinations from Destination Service you need to do the following steps: 

Reviewer: ```suggestion
 In order to fetch all destinations from Destination Service you need to do the following steps: 
```

>  In order to fetch all destinations from Destination Service you need to do the following steps: 

Reviewer 3: ```suggestion
In order to fetch all destinations from the Destination Service you need to perform the following steps: 
```

 1) Invoke the `ScpCfDestinationLoader` with the `DestinationService` and `XsuaaService` instances.

```java

>  final ScpCfDestinationLoader destinationLoader = new ScpCfDestinationLoader(destinationService, xsuaaService);

Reviewer 2: You are referencing a package-private constructor. Instead use:
```suggestion
 final ScpCfDestinationLoader destinationLoader = new ScpCfDestinationLoader();
```

```

>  2) Build `DestinationOptions` object based on the retrieval strategy for your particular use case. Retieval strategy can be `ALWAYS_SUBSCRIBER`, `ALWAYS_PROVIDER` or `SUBSCRIBER_THEN_PROVIDER`. Refer the code below for better understanding:  

Author: I would like to know if this explanation on DestinationOptions object is sufficient? (i.e we need it here only to configure different retrieval strategies.) or I am missing something here?

Reviewer 3: Honestly I am quite surprised that:

a) the default is always subscriber (that is not consistent with the default of `tryGetDestination` which is subscriber then provider)

and b) that we need this at all. I would have expected just `tryGetAllDestinations()` or `tryGetAllDestinations(RetrievalStrategy)`.

Author: > a) the default is always subscriber (that is not consistent with the default of tryGetDestination which is subscriber then provider)

I guess the default for `tryGetDestination ` is also always subscriber. I understood this by Line140 in `ScpCfDestinationLoader`.

> and b) that we need this at all. I would have expected just tryGetAllDestinations() or tryGetAllDestinations(RetrievalStrategy).

 I think we need this to set the Retrieval strategy. Hence,  you are right the API could have been just `tryGetAllDestinations(RetrievalStrategy).` . 

I would like to know what Alex think about this.

Reviewer 2: > I guess the default for `tryGetDestination ` is also always subscriber. I understood this by Line140 in `ScpCfDestinationLoader`.

I would second that. This seems to be the case [since 3.0.0](https://github.wdf.sap.corp/MA/sdk/blob/rel/3.0.0/cloudplatform/cloudplatform-connectivity-scp-cf/src/main/java/com/sap/cloud/sdk/cloudplatform/connectivity/ScpCfDestinationLoader.java#L103).

> > b) that we need this at all. I would have expected just tryGetAllDestinations() or tryGetAllDestinations(RetrievalStrategy).
> 
> I think we need this to set the Retrieval strategy. Hence, you are right the API could have been just `tryGetAllDestinations(RetrievalStrategy).`

Please note the rest of the API. The `RetrievalStrategy` is only exposed through `DestinationOptions`. You may oppose this decision now, but it was made from the beginning and changing that is not possible without breaking everything.

The loader classes, i.e. `ScpCfDestinationLoader`, are made public for **advanced** users. Convenience methods are to be exposed through `DestinationAccessor`. Since this feature is only available on CF, we cannot expose it on that general class. Hence we are talking about an advanced feature.

If you are challenging the API, please create a team-internal follow-up issue. Also include Chris, one major author of the Destinations API.

Author: I am ok with the current API.

>  2) Build `DestinationOptions` object based on the retrieval strategy for your particular use case. Retieval strategy can be `ALWAYS_SUBSCRIBER`, `ALWAYS_PROVIDER` or `SUBSCRIBER_THEN_PROVIDER`. Refer the code below for better understanding:  

Reviewer 3: ```suggestion
 2) Build a `DestinationOptions` object based on the retrieval strategy for your particular use case. The Retieval strategy can be `ALWAYS_SUBSCRIBER`, `ALWAYS_PROVIDER` or `SUBSCRIBER_THEN_PROVIDER`. E.g. for `SUBSCRIBER_THEN_PROVIDER` use:
```

>  2) Build a `DestinationOptions` object based on the retrieval strategy for your particular use case. The Retieval strategy can be `ALWAYS_SUBSCRIBER`, `ALWAYS_PROVIDER` or `SUBSCRIBER_THEN_PROVIDER`. E.g. for `SUBSCRIBER_THEN_PROVIDER` use:

Reviewer 2: ```suggestion
 2) For the upcoming call build a `DestinationOptions` object. It contains the configuration on how the destination service is being queried. If you have a simple application without provider/subscriber setup, then your initial configuration may look like this:
    ```java
    // in case of Subscriber Tenant
    final DestinationOptions options = DestinationOptions.builder().build();
    ```
  If you have a provider/subscriber setup, a retrieval strategy must be chosen according to your particular use case. The retrieval strategy can be `ALWAYS_SUBSCRIBER`, `ALWAYS_PROVIDER` or `SUBSCRIBER_THEN_PROVIDER`. E.g. for `SUBSCRIBER_THEN_PROVIDER` use:
```

 ```java
 // in case of Subscriber Tenant
 final DestinationOptions options =
            DestinationOptions
                .builder()
                .augmentBuilder(
                    ScpCfDestinationOptionsAugmenter.augmenter().retrievalStrategy(
                        ScpCfDestinationRetrievalStrategy.SUBSCRIBER_THEN_PROVIDER))
                .build();;
```

3) Finally, make call to `tryGetAllDestinations` and pass the `DestinationOptions` object.

```java
 final Try<Iterable<ScpCfDestination>> destinations = destinationLoader.tryGetAllDestinations(options);
```

> The method `tryGetAllDestinations` requests the [Destination Service API](https://api.sap.com/api/SAP_CP_CF_Connectivity_Destination/overview) and retrieves all the destinations available at the service instance and sub account level. In case there is a destination available on both the levels with the same name, then this method prioritize the destination at the service instance level.

Reviewer: ```suggestion
The method `tryGetAllDestinations` requests the [Destination Service API](https://api.sap.com/api/SAP_CP_CF_Connectivity_Destination/overview) and retrieves all the destinations available at the service instance and sub-account level. In case there is a destination available on both the levels with the same name, then this method prioritizes the destination at the service instance level.
```

Author: Good catch:)

> The method `tryGetAllDestinations` requests the [Destination Service API](https://api.sap.com/api/SAP_CP_CF_Connectivity_Destination/overview) and retrieves all the destinations available at the service instance and sub-account level. In case there is a destination available on both the levels with the same name, then this method prioritizes the destination at the service instance level.

Reviewer 3: ```suggestion
The method `tryGetAllDestinations` queries the [Destination Service API](https://api.sap.com/api/SAP_CP_CF_Connectivity_Destination/overview) and retrieves all the destinations available at the service instance and sub-account level. In case there is a destination available on both the levels with the same name, then this method prioritizes the destination at the service instance level.
```

This concept is integrated with the [Destination Service](https://help.sap.com/viewer/cca91383641e40ffbe03bdc78f00f681/Cloud/en-US/7e306250e08340f89d6c103e28840f30.html) that is available on SAP Cloud Platform. If the application has a service binding to this service in place the SDK will provide access to these destinations.

> ## Get a List of all destinations from Destination Service on Cloud Foundry  ##

Reviewer 3: IMO this should be placed below the other sections. The reason is simply that it is the less common, more advanced use case that fewer people will need. It should be clear that the usual way to retrieve destinations is through the `DestinationAccessor`.

Reviewer 3: ```suggestion
## Retrieve all destinations from the Destination Service on Cloud Foundry  ##
```

Author: ok I agree. i will place it below.

 
In order to fetch all destinations from the Destination Service you need to make a call to `tryGetAllDestinations`.The method queries the [Destination Service API](https://api.sap.com/api/SAP_CP_CF_Connectivity_Destination/overview) and retrieves all the destinations available at the service instance and sub-account level. In case there is a destination available on both the levels with the same name, then this method prioritizes the destination at the service instance level.

Below is the sample call to `tryGetAllDestinations`:

```java
 final Try<Iterable<ScpCfDestination>> destinations = destinationLoader.tryGetAllDestinations(options);
```

> 

Reviewer 3: I think one newline is enough. That helps keeping the document short.

>  In the above call, `destinationLoader` is the instance of `ScpCfDestinationLoader`.

Reviewer 3: ```suggestion
In the above call `destinationLoader` needs to be an instance of `ScpCfDestinationLoader`.
```
